### PR TITLE
Device-side testing.

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michal Zientkiewicz <michalz@nvidia.com>

### CODE REVIEW

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

File: dali/test/device_test_test.cu [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+TEST(DeviceTest, DeviceSideFailure) {
+  int num = 0;
+  EXPECT_NONFATAL_FAILURE({
+    DEVICE_TEST_CASE_BODY(DeviceTest, DeviceSideFailure, dim3(1), dim3(2))
+    num = host_status.num_messages;
+  }, "There were errors in device code");
+  EXPECT_EQ(num, 21);  // threadIdx.x == 0 succeeds once
+  EXPECT_EQ(cudaGetLastError(), cudaSuccess);
```
Comment (jantonguirao) [inline dali/test/device_test_test.cu]: why isn't there an error?

File: include/dali/core/cuda_utils.h [hunk 2]
```diff
@@ -49,10 +49,54 @@ __device__ inline float16 StaticCastGpu(float val) {
...
 
-// Starts profiling DALI
-inline void DALIProfilerStart() { cudaProfilerStart(); }
+// allocator using device-side malloc/free
 
-inline void DALIProfilerStop() { cudaProfilerStop(); }
+template <typename T>
+struct device_side_allocator {
+  static __device__ T *allocate(size_t count) {
+    return static_cast<T *>(malloc(count * sizeof(T)));
```
Comment (jantonguirao) [inline include/dali/core/cuda_utils.h]: shouldn't it be cudaMalloc for GPU?

File: dali/test/device_test_test.cu (hunk 1 above)
Comment (mzient) [reply dali/test/device_test_test.cu]: I cannot raise a CUDA runtime error programmatically. If I raise an assert in a kernel, then the cuda context is no longer usable (it's akin to C-style `assert` - it assumes that program is broken should not be terminated).

File: include/dali/core/cuda_utils.h (hunk 2 above)
Comment (mzient) [reply include/dali/core/cuda_utils.h]: No, it's device-side allocator. On device, there are `malloc` and `free` functions which allocate global memory unusable outside device code. Unlike memory allocated with `cudaMalloc`, this memory cannot be used by runtime APIs, such as `cudaMemcpy` and cannot be freed by host code. The amount of this memory is very limited, too - 8MB by default, AFAIK.

Comment (jantonguirao) [reply include/dali/core/cuda_utils.h]: Got it, thanks for the explanation!

**Pull request merged**